File: backend/src/main/java/lan/dk/podcastserver/entity/Status.java
package lan.dk.podcastserver.entity;

import com.fasterxml.jackson.annotation.JsonCreator;
import io.vavr.collection.Set;
import io.vavr.control.Option;

import static io.vavr.API.Set;

public enum Status {
    NOT_DOWNLOADED,
    STARTED,
    PAUSED,
    DELETED,
    STOPPED,
    FAILED,
    FINISH;

    /**
     * Created by kevin on 18/02/15.
     */
    private static final Set<Status> _values = Set(Status.values());

    @JsonCreator
    public static Status of(String v) {
        return Status.from(v).getOrElseThrow(() -> new IllegalArgumentException("No enum constant Status." + v));
    }

    public static Option<Status> from(String v) {
        return _values.find(s -> s.toString().equals(v));
    }
    
}

File: backend/src/main/kotlin/com/github/davinkevin/podcastserver/entity/Status.kt
package com.github.davinkevin.podcastserver.entity

import com.fasterxml.jackson.annotation.JsonCreator
import com.github.davinkevin.podcastserver.database.enums.ItemStatus

enum class Status {
    NOT_DOWNLOADED,
    STARTED,
    PAUSED,
    DELETED,
    STOPPED,
    FAILED,
    FINISH;

    companion object {

        private val values = setOf(*values())

        @JsonCreator
        @JvmStatic
        fun of(v: String): Status {
            return values
                    .firstOrNull { s -> s.toString() == v }
                    ?: throw IllegalArgumentException("No enum constant Status.$v")
        }

    }
}

fun Status.toDb(): ItemStatus = when(this) {
    Status.NOT_DOWNLOADED -> ItemStatus.NOT_DOWNLOADED
    Status.STARTED -> ItemStatus.STARTED
    Status.PAUSED -> ItemStatus.PAUSED
    Status.DELETED -> ItemStatus.DELETED
    Status.STOPPED -> ItemStatus.STOPPED
    Status.FAILED -> ItemStatus.FAILED
    Status.FINISH -> ItemStatus.FINISH
}

fun ItemStatus.fromDb(): Status = when(this) {
    ItemStatus.NOT_DOWNLOADED -> Status.NOT_DOWNLOADED
    ItemStatus.STARTED -> Status.STARTED
    ItemStatus.PAUSED -> Status.PAUSED
    ItemStatus.DELETED -> Status.DELETED
    ItemStatus.STOPPED -> Status.STOPPED
    ItemStatus.FAILED -> Status.FAILED
    ItemStatus.FINISH -> Status.FINISH
}
